django_toolkit/fallbacks/circuit_breaker/__init__.py

```diff
@@ -0,0 +1,116 @@
+# -*- coding: utf-8 -*-
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class CircuitBreaker:
+
+    def __init__(
+        self,
+        rule,
+        cache,
+        failure_exception,
+        failure_timeout=None,
+        circuit_timeout=None,
+        catch_exceptions=None,
+    ):
+        self.rule = rule
+        self.cache = cache
+        self.failure_timeout = failure_timeout
+        self.circuit_timeout = circuit_timeout
+        self.circuit_cache_key = 'circuit_{}'.format(rule.failure_cache_key)
+        self.failure_exception = failure_exception
+        self.catch_exceptions = catch_exceptions or (Exception,)
+        self.failure_cache_key = self.rule.failure_cache_key
+        self.request_cache_key = getattr(self.rule, 'request_cache_key', '')
+
+    @property
+    def is_circuit_open(self):
+        return self.cache.get(self.circuit_cache_key) or False
+
+    @property
+    def total_failures(self):
+        return self.cache.get(self.rule.failure_cache_key) or 0
+
+    @property
+    def total_requests(self):
+        return self.cache.get(self.request_cache_key) or 0
+
+    @property
+    def percentage_failures(self):
+        if self.total_requests > 0:
+            return (self.total_failures * 100) / self.total_requests
+        return 0
+
+    def open_circuit(self):
+        self.cache.set(self.circuit_cache_key, True, self.circuit_timeout)
+
+        # Delete the cache key to mitigate multiple sequentials openings
+        # when a key is created accidentally without timeout (from an incr
+        # operation)
+        self.cache.delete(self.rule.failure_cache_key)
+        self.cache.delete(self.request_cache_key)
+
+        logger.critical(
+            'Open circuit for {failure_cache_key} {cicuit_cache_key}'.format(
+                failure_cache_key=self.failure_cache_key,
+                cicuit_cache_key=self.circuit_cache_key
+            )
+        )
+
+    def __enter__(self):
+        if self.is_circuit_open:
+            raise self.failure_exception
+
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self._increase_request_count()
+        if exc_type in self.catch_exceptions:
+            if self.is_circuit_open:
+                raise self.failure_exception
+
+            self._increase_failure_count()
+
+            if self.rule.should_open_circuit(
+                total_failures=self.total_failures,
+                percentage_failures=self.percentage_failures
+            ):
```

Comment:
What if `should_open_circuit` accept `total_failures` and `total_requests` parameters and only `PercentageFailuresRule` use it to calculate `percentage_failures`?

Reply:
I don't think so, as `percentage_failures` is only used by `PercentageFailuresRule`.

Reply:
now all makes sense